src/wp-admin/includes/class-wp-automatic-updater.php

```diff
@@ -1551,4 +1678,84 @@ protected function send_debug_email() {
 
 		wp_mail( $email['to'], wp_specialchars_decode( $email['subject'] ), $email['body'], $email['headers'] );
 	}
+
+	/**
+	 * Performs a loopback request to check for potential fatal errors.
+	 *
+	 * Fatal errors cannot be detected unless maintenance mode is enabled.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @global int $upgrading The Unix timestamp marking when upgrading WordPress began.
+	 *
+	 * @return bool Whether a fatal error was detected.
+	 */
+	protected function has_fatal_error() {
+		global $upgrading;
+
+		$maintenance_file = ABSPATH . '.maintenance';
+		if ( ! file_exists( $maintenance_file ) ) {
+			return false;
+		}
+
+		require $maintenance_file;
+		if ( ! is_int( $upgrading ) ) {
+			return false;
+		}
+
+		$scrape_key   = md5( $upgrading );
+		$scrape_nonce = (string) $upgrading;
+		$transient    = 'scrape_key_' . $scrape_key;
+		set_transient( $transient, $scrape_nonce, 30 );
+
+		$cookies       = wp_unslash( $_COOKIE );
+		$scrape_params = array(
+			'wp_scrape_key'   => $scrape_key,
+			'wp_scrape_nonce' => $scrape_nonce,
+		);
+		$headers       = array(
+			'Cache-Control' => 'no-cache',
+		);
+
+		/** This filter is documented in wp-includes/class-wp-http-streams.php */
+		$sslverify = apply_filters( 'https_local_ssl_verify', false );
+
+		// Include Basic auth in the loopback request.
+		if ( isset( $_SERVER['PHP_AUTH_USER'] ) && isset( $_SERVER['PHP_AUTH_PW'] ) ) {
+			$headers['Authorization'] = 'Basic ' . base64_encode( wp_unslash( $_SERVER['PHP_AUTH_USER'] ) . ':' . wp_unslash( $_SERVER['PHP_AUTH_PW'] ) );
+		}
+
+		// Time to wait for loopback request to finish.
+		$timeout = 50; // 50 seconds.
+
+		error_log( '    Scraping home page...' );
+
+		$needle_start = "###### wp_scraping_result_start:$scrape_key ######";
+		$needle_end   = "###### wp_scraping_result_end:$scrape_key ######";
+		$url          = add_query_arg( $scrape_params, home_url( '/' ) );
```

Comment:
Can we make the `$url`  filterable? Many plugins don't affect the frontpage (or shared lifecycle) and won't be caught by a naive check. Adding a filter here would enable a guardrail for plugin developers to endure a relevant rollback check. E.g.

- A utility plugin checking it's Admin screen loads.
- A CPT plugin checking the CPT archive doesn't fatal.
- A large ecosystem plugin checking a custom endpoint that runs a series of post-update health checks.

Reply:
> Can we make the $url filterable?
That's actually something we considered during the feature's earlier development. It was removed as we wanted to start with the core feature and iterate as requests were made. I'm happy to add the filter, but I'll leave this conversation open for a bit first to see if there is more support for it, or possibly some opposition to it at this time.

>- A utility plugin checking it's Admin screen loads.
> - A large ecosystem plugin checking a custom endpoint that runs a series of post-update health checks.

Loopback requests for admin will require authentication, so this is likely not possible with the current setup (and I _believe_ that's the case with the Plugin File Editor and Theme File Editor too). It's certainly something we could look at in a future iteration, though I think it might require somewhat of a deep-dive to look at implementation options.

> - A CPT plugin checking the CPT archive doesn't fatal.

Yep, that would be achievable with a filter, assuming the CPT archive is a public page (usually is, but some may be "members-only" areas).

Reply:
Thanks @costdev !

> I'm happy to add the filter, but I'll leave this conversation open for a bit first to see if there is more support for it, or possibly some opposition to it at this time.

I'll share in `#outreach` to get some more voices to weigh in 🤞 

> A utility plugin checking it's Admin screen loads.
A large ecosystem plugin checking a custom endpoint that runs a series of post-update health checks.
Loopback requests for admin will require authentication, so this is likely not possible with the current setup (and I believe that's the case with the Plugin File Editor and Theme File Editor too).

Great catch! While it wouldnt be too complicated to make the rest of the `wp_remote_get()` params filterable, that definitely is overkill for an initial implementation. Besides, utility plugins could also take the approach of using a custom endpoint, even if that's a bit more work than just checking an admin screen they've already developed/hooked into.

Reply:
The goal is to prevent a site from a PHP fatal after a plugin update. That typically means checking the homepage. If some custom endpoint results in a PHP fatal will most users even see that?

Reply:
@afragen touches on something here: Our main targets during development were fatal errors that would completely render a site unusable, rather than say, checking a plugin's functionality. Not that a single page type fataling isn't bad - of course it is - but our main focus at this stage has been ones where the homepage can't load, much like the Plugin File Editor and Theme File Editor, and the file edited in those editors could be for any page/template/endpoint.

Also note that a filtered URL would _replace_ the URL to be scraped, rather than add another URL to scrape. That means a plugin's URL may work, while the homepage doesn't. We can't scrape multiple URLs because that would cause an unknown delay to every update.

Where I'm at currently (pending further discussion): For consistency with the Plugin File Editor and Theme File Editor, as well as ensuring the homepage is protected, I'm leaning towards keeping the homepage as the only URL that's scraped, at least for this initial version of the feature. Enhancing too early can risk undermining the feature, and since this is a stability feature, that'd be counter-productive.

Reply:
@afragen 

>The goal is to prevent a site from a PHP fatal after a plugin update. That typically means checking the homepage.If some custom endpoint results in a PHP fatal will most users even see that?

"Typically" or currently?
If Co-Authors Plus breaks on a product site, it's the blog page that will _always_ fatal, while the home page only will fatal if theres a recent posts section.
Same thing with WooCommerce+ the Shop endpoint. An extension like Woo Payments or  WooCommerce [x] Gateway would naively affect the checkout page, and more likely fatal during the checkout _process_, which is where a custom health check endpoint could be used.

(@costdev , I believe most would consider the above examples as "fatal errors that would completely render a site unusable", from either an end user or developer perspective)

> Where I'm at currently (pending further discussion): For consistency with the Plugin File Editor and Theme File Editor, as well as ensuring the homepage is protected, I'm leaning towards keeping the homepage as the only URL that's scraped, at least for this initial version of the feature.

Using the Plugin/Theme File Editor as the measure of consistency seems like the wrong benchmark IMO.
- From an engineering POV: at a glance, it doesn't look like this feature is sharing code with the Plugin/Theme checker, so there's no technological need for consistency.
- From an end-user POV, the use cases are significantly different:
  - If you're actively making changes to an existing plugin, then you (ostensibly) know how to test what what you're changing. In this scenario, the Fatal Error checker is a quick, extra safeguard to make sure you didn't  accidentally do something obviously broken (that would otherwise have been caught by your IDE/CI if you weren't making code changes via the browser). More targeted health checks are wholly additive, don't affect the goal of the feature, and offer limited benefit.
  - The goal of Plugin Rollbacks (as I understand it) is provide some form of protection against incompatible code changes by a third party, and even more so in an auto-update scenario where youre not even aware of these changes until after the fact. As such a naive homepage check doesn't provide an additive guard, but rather a limited and often _ineffective_ guard.
  In the best case scenario, the feature limitations are clearly communicated and users stay opted-out of Auto-Updates for most of their plugins; worst case, the false expectations have users turning on Auto-Updates for more plugins, they still get updates that break their site, and are even more trepidatious about Auto Updates in the future.
  
> Enhancing too early can risk undermining the feature, and since this is a stability feature, that'd be counter-productive.



Per the above, I believe the risk here is reversed. From a user perspective, an inneffectice guard is more "unstable" (in context: "unreliable") than.... What exactly?

As a reminder, we're talking about plugins additively being able to provide a different string URI (that's more relevant than the homepage) to loopback instead of the default hard-coded `/`. The minimal API footprint is the URI string and the plugin identifier. What are some cases you worried about that would undermine the feature?

Reply:
_Before reading_: I'm on mobile and between tasks, so hopefully my brain -> keyboard coordination holds up here :sweat_smile:

> I believe most would consider the above examples as "fatal errors that would completely render a site unusable", from either an end user or developer perspective

Though I'm not talking about a subjective interpretation of a site being unusable, I might even agree with every example you've given. I'm talking about when a site's homepage doesn't load, which may indicate a more fundamental browsability issue. i.e. If a homepage fatals, the site is unusable by everyone trying to view that site's top-level URL, the main arrival point and "base", and there's nowhere to go from there. If a sub-page breaks, the site is still browsable from its top-level URL. A sub-page may work while the homepage doesn't, but who can reach the sub-page to find out?

For example, WooCommerce could choose one of many endpoint URLs where something could go wrong and it's critical to the plugin's functionality. If it picked one, I imagine it would be the shop page as its main entry point (i.e. no point verifying a later point you can't reach because the shop page isn't even working). It would be limited and possibly ineffective, but it'd at least be trying to protect users at its entry point. That's akin to what we're doing here, as WordPress' entrypoint. Even if other URL checks were desired, a check on the shop page shouldn't be replaced, and that's again where I'm hesitant to introduce a filter that would replace our check on the homepage; a site's main entry point.

You mentioned about a plugin that might only affect a blog page, though for example, any plugin intended to have its "own space" may hook `init`, and contain a conflict and fatal error in the callback that runs on every page, thus objectively rendering _the whole site_ unusable. That's the kind of fundamental issue I mean.

---

> From an engineering POV: at a glance, it doesn't look like this feature is sharing code with the Plugin/Theme checker, so there's no technological need for consistency.

This isn't my response but just an FYI: The feature uses the same (not a copy) scraping code as the Plugin File Editor and Theme File Editor. While the request is its own in `::has_fatal_error()`, the actual scraping is the same code.

While we're aiming for consistency at an engineering level too, the consistency I referred to earlier isn't necessarily at an engineering level, but that the homepage is checked, and also the thought process.

From [the commit](https://core.trac.wordpress.org/changeset/41721) that introduced these editors:
> Introduce sandboxed PHP file edits for active themes, preventing accidental whitescreening of a user's site when introducing a fatal error.
> After writing a change to a PHP file for an active theme or plugin, perform loopback requests on the file editor admin screens and the homepage to check for fatal errors. If a fatal error is encountered, roll back the edited file and display the error to the user to fix and try again.

And from the discussion on one of the [related tickets](https://core.trac.wordpress.org/ticket/21622#comment:18):

> we can at least make sure the front page still loads to mitigate some forms of breakage is my thinking.

---

> If you're actively making changes to an existing plugin, then you (ostensibly) know how to test what what you're changing. In this scenario, the Fatal Error checker is a quick, extra safeguard to make sure you didn't accidentally do something obviously broken (that would otherwise have been caught by your IDE/CI if you weren't making code changes via the browser). More targeted health checks are wholly additive, don't affect the goal of the feature, and offer limited benefit.

The fatal error checker for the editors is quite critical. It prevents you editing a file such that you may not be able to access the same admin screen to repair the issue, and performs an additional check to the homepage such that the site is still browsable. These editors are often used by users of all ranges, from experienced developers with no FTP access, to site owners with no code experience who are using snippets they found on a blog post, so whether they know exactly how to test (fully) the potential impact of their changes is 50/50.

> The goal of Plugin Rollbacks (as I understand it) is provide some form of protection against incompatible code changes by a third party, and even more so in an auto-update scenario where youre not even aware of these changes until after the fact. As such a naive homepage check doesn't provide an additive guard, but rather a limited and often ineffective guard.
In the best case scenario, the feature limitations are clearly communicated and users stay opted-out of Auto-Updates for most of their plugins; worst case, the false expectations have users turning on Auto-Updates for more plugins, they still get updates that break their site, and are even more trepidatious about Auto Updates in the future.


As demonstrated earlier in this thread, I do see the value in a more targeted check by plugins, but not as a replacement for a homepage check. I can hope a filter would only be used when a plugin author is positive that their plugin's code doesn't run on the homepage, or during any bootstrapping/init, but historically, intended usage and actual usage differ and have a bunch of use cases in between. What we aim to guarantee with Rollback Auto-Update, at least in this initial version, is that an automatic update will not present visitors to your homepage with a WSOD that may indicate issues that could be present on every other page of your site.

As an aside and a personal preference, if it's okay with you I'd prefer not talk in terms of best/worst case scenarios while discussing a disagreement, as it can often result in a back and forth that eventually (even surprisingly) becomes emotive, with people nit-picking at each other's speculation that's required for best/worst cases, and would rather continue our discussion as we have earlier in the thread. It just tends to be a better discussion, at least in my opinion/experience.

> Per the above, I believe the risk here is reversed. From a user perspective, an inneffectice guard is more "unstable" (in context: "unreliable") than.... What exactly?

Unfortunately (and I hate it), we can never be 100% reliable against fatal errors as we can't execute all code paths in a plugin. What we can do is make sure that the site still loads more than a WSOD at its entrypoint. IMO, that should be a _base_ check, with any others being additional scrapes - and that's where the problem with a filter lies for me, either it replaces the homepage check, or it's additional and could put the whole update process at risk of timeout, etc.

> As a reminder, we're talking about plugins additively being able to provide a different string URI (that's more relevant than the homepage) to loopback instead of the default hard-coded /. The minimal API footprint is the URI string and the plugin identifier. What are some cases you worried about that would undermine the feature?

I mentioned above about an `init` hook's callback containing a conflict that occurs on every page, and a plugin author can't reasonably be expected to have tested against all plugins for a conflict.

Another case would be that a hook that should run on every page assumes the availability of a plugin's functionality, and it only worked when a plugin author tested it because of how they'd configured their site. In other words, a case where "happy path"-only testing strikes again.

---

With all the above said, I like the idea of a plugin being able to check that a particular page can load, but not as a replacement of the homepage check. I just can't see a way to accomplish that without impacting the update process' execution. That may just be a limitation of my imagination at the moment, I don't know, but that's why I want to keep discussing this :slightly_smiling_face:

Reply:
Thanks @costdev so much for the attention and detail - especially from mobile!

Before anything else:
> As an aside and a personal preference, if it's okay with you I'd prefer not talk in terms of best/worst case scenarios while discussing a disagreement, as it can often result in a back and forth that eventually (even surprisingly) becomes emotive.

This is a lovely and appreciated tip. Thank you 💙 

> I'm talking about when a site's homepage doesn't load, which may indicate a more fundamental browsability issue. i.e. If a homepage fatals, the site is unusable by everyone trying to view that site's top-level URL, the main arrival point and "base", and there's nowhere to go from there. If a sub-page breaks, the site is still browsable from its top-level URL. A sub-page may work while the homepage doesn't, but who can reach the sub-page to find out?

_This_ is the important context I was missing that brings everything else into focus. I was looking at this feature as a guard against plugins breaking _on_ the site, and not plugins breaking _access to_ the site. From the latter perspective, we indeed wouldn't want to replace the `home_url` check, and while it doesn't address the larger concerns about auto-update stability that sort of work makes much more sense as possible enhancements only after this more fundamental check is in place 👍

Responding to a couple of your individual points for future reference (either others joining this discussion, or prior art review when its time to start talking future possible enhancements).

---

> The fatal error checker for the editors is quite critical. It prevents you editing a file such that you may not be able to access the same admin screen to repair the issue, and performs an additional check to the homepage such that the site is still browsable.

Understood. It was for this reason I suggested the use case and thus the guard required are different. Users aren't editing code when updating a plugin, or taking _any_ manual actions during Auto-Updates. Ensuring users don't lock themselves out of being able to revert the change doesn't apply (unless/until a rollback GUI is introduced), and from a _broken plugin_ POV the homepage check isn't a strong indicator of anything. From a _site access_ POV, that homepage check still resonates.

> I do see the value in a more targeted check by plugins, but not as a replacement for a homepage check. I can hope a filter would only be used when a plugin author is positive that their plugin's code doesn't run on the homepage, or during any bootstrapping/init, but historically, intended usage and actual usage differ and have a bunch of use cases in between.

Agreed - seems like future enhancements to the feature would be better served by an action that allows plugins to run their own additional health check, without replacing the home_url access check.